feat(elixir): refactor secure channel

polvorin opened the pull request:
* Unify secure channel with identity secure channels  (-> all secure channels are "identity" secure channels)
* Use ockam worker (`AsymetricWorker` to get two different addresses) for implementation
* Use a single state machine,  prepare to adapt to changes on https://github.com/build-trust/ockam/pull/4897
* Add couple missing tests  (secure channels peers get terminated if peer gracefully terminate it,   tests around metric collection, etc)
* Move particularities of our implementation  (bare-encoding of binary messages,  routes, etc) out of the XX handshake module and into the secure channel itself
```
               ┌─────────────────┐                          ┌─────────────────┐
               │                 │                          │                 │
           ┌───┴──────┐    ┌─────┴────┐                 ┌───┴──────┐    ┌─────┴────┐
Plaintext  │ Address  │    │Inner     │                 │Inner     │    │ Address  │  Plaintext
  ◄──────► │          │    │Address   │ ◄────[...] ───► │Address   │    │          │ ◄──────►
           └───┬──────┘    └─────┬────┘                 └───┬──────┘    └─────┬────┘
               │  SecureChannel  │       Ciphertext         │   SecureChannel │
               └─────────────────┘                          └─────────────────┘
```

polvorin committed:
feat(elixir): refactor secure channel

unify secure channel with identity secure channels,
use ockam worker for implementation.

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
+          | {:idle_timeout, non_neg_integer() | :infinity}
+
+  # Handshaking | IdentityExchange | Established are the different states in which
+  # the connection can be.   We don't use a proper fsm here as not directly supported
```
This is great context - might be good for inclusion in the moduledoc?

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
-    channel |> Node.whereis() |> GenStateMachine.call(:established?)
+  @type encryption_options :: [{:vault, Ockam.Vault.t()}, {:static_keypair, reference()}]
+  @type authorization :: list() | map()
+  @type trust_policies :: list()
```
Are we able to be more specific about the inner type, such as `list(binary())` or whatever would be correct? Or is this already the appropriate precision?

polvorin replied:
:+1:   yes we should be able to be more specific,  will take a look at it.    I let at `list` and `map` when doing the refactor as as wasn't sure what exactly were the valid values for trust_policies or authorizations

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
+    field(:channel_state, Handshaking.t() | IdentityExchange.t() | Established.t())
+  end
+
+  @handshake_timeout 20_000
```
Deliberate that this differs from the outer timeout on the `defdelegate`?

polvorin replied:
no.    will fix!

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
+  end
+
+  @spec start_link_channel([initiator_opt()], non_neg_integer) :: :ok | {:error, term()}
+  def start_link_channel(opts, handshake_timeout \\ 3000) do
```
👀 Third unique value for a timeout I've seen, probably correct but just making sure.

polvorin replied:
:grimacing:   and this one is wrong

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
-
-      error ->
-        error
+  # Deprecated, use start_link_channel/2
```
Optional, as it produces compilation warnings, but good for consumers' feedback IMO:
```suggestion
  @deprecated "use start_link_channel/2"
```

polvorin replied:
Yes letting it as a comment for now.  There is code using this and the behavior is slightly different (I copied the deprecated comment from the current version)

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
-    case established?(address) do
-      false ->
-        :timer.sleep(interval)
-        wait_for_channel(pid, address, :infinity, interval)
+  @doc """
+  Stop secure channel and it's remote endpoint
+  """
+  def disconnect(worker) do
+    # TODO: a better solution is needed, this works
+    # in a best-effort manner as long as disconnect() is being called
+    Ockam.Worker.call(worker, :disconnect)
+  end
 
-      true ->
-        {:ok, pid, address}
-    end
+  def role(worker) do
+    Ockam.Worker.call(worker, :role)
   end
 
-  @doc false
-  def start_link(options) when is_list(options) do
-    address = Keyword.get(options, :address, Node.get_random_unregistered_address())
-    options = Keyword.put(options, :address, address)
+  def established?(worker) do
+    Ockam.Worker.call(worker, :established?)
+  end
```
Optional, if we're not going to supply docstrings for some of these, and they're just one-to-one wrappers of `call`, I'd consider metaprogramming them for brevity:

```
for msg <- ~w(disconnect established? get_remote_identity get_remote_identity_id role)a do # and so on
  def unquote(msg)(worker), do: Ockam.Worker.call(worker, unquote(msg))
end

polvorin replied:
will pass on that :),   having type specs and docstrings will be helpful, we should start adding those at least on this kind of more user-facing places.

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
+
+  defp new_xx(vault, opts) do
+    m1_payload = m2_payload = m3_payload = ""
+    {:ok, e_key_handle} = Ockam.Vault.secret_generate(vault, type: :curve25519)
+    {:ok, e_public_key} = Vault.secret_publickey_get(vault, e_key_handle)
```
Nit, but I'd like to see both calls either aliased or fully-qualified as appropriate.

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
-          {:next_state, state, data} ->
-            {:next_state, state, data}
+  defp identity_from_opts(options) do
+    identity_module = Keyword.get(options, :identity_module, Identity.default_implementation())
```
```suggestion
    identity_module = Keyword.get_lazy(options, :identity_module, &Identity.default_implementation/0)
```

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
+      end
+    else
+      # The message couldn't be decrypted.  State remains unchanged
+      _error -> {:ok, state}
```
Should this be logged, metered with telemetry, or anything similar to give us some insight later?

polvorin replied:
added it to logs

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
+    worker_opts =
+      opts
+      |> Keyword.drop([:address, :authorization])
+      |> Enum.map(fn
```
Nit: Almost zero practical difference in the outcome but slightly more intention-revealing IMO. The major distinction is it de-duplicates on purpose, which I'm not sure is benign/good/bad for us here.
```suggestion
      |> Keyword.new(fn
```

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,611 @@
-      unexpected_protocol ->
-        {:error, {:unexpected_key_establishment_protocol, unexpected_protocol}}
-    end
+    Keyword.merge(Keyword.take(opts, [:address, :authorization]),
```
This is nicely clear as-written, but you could also begin the function with `{opts, worker_opts} = Keyword.split(opts, [:address, authorization])` and go from there.

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/key_establishment_protocol/xx/protocol.ex:
```diff
@@ -119,55 +145,22 @@ defmodule Ockam.SecureChannel.KeyEstablishmentProtocol.XX.Protocol do
-    decoder = String.to_existing_atom("decode_" <> Atom.to_string(message_name))
-    return_value = apply(__MODULE__, decoder, [message, protocol_state])
+  def encode(:message1, %{e: e, payloads: payloads} = state) do
+    payload = Map.get(payloads, :message1, "")
```
Will we eventually want to remove the default value here after the three-message work, or is always reasonable to have a fallback to empty string?

polvorin replied:
good question.   Even after the upcoming changes,  the _first_ handshake message won't carry any payload in our protocol.  So this could had even be hardwired to be the an empty binary. 
_But_   there are existing test vectors we use for unit tests of the noise handshake that _do_ have payloads..  so that's right now the only value of keeping it open (that we can validate existing test vectors for noise,  not only ones explicitly written by us)

shanesveller commented on implementations/elixir/ockam/ockam/lib/ockam/worker/authorization.ex:
```diff
@@ -119,7 +119,7 @@ defmodule Ockam.Worker.Authorization do
   """
   def from_secure_channel(prev \\ :ok, message, _state) do
-    Logger.debug("check is secure #{inspect(message)}")
+    Logger.debug("check from secure channel #{inspect(message)}")
```
I realize this is no worse than the existing impl, but can we be any more conservative about limiting the inspect output? Do messages have a useful Inspect protocol definition?

polvorin replied:
some explicit textual representation would be nice yes.  
So far we are logging the message this way in lots of places,  it's still "not-to-bad"  cause the things most likely to be of interest  (from/to  route,  and metadata)  are short enough and have okish text representation already.
Message' body is truncated

shanesveller commented on implementations/elixir/ockam/ockam/mix.exs:
```diff
@@ -14,7 +14,7 @@ defmodule Ockam.MixProject do
       elixir: @elixir_requirement,
       consolidate_protocols: Mix.env() != :test,
-      elixirc_options: [warnings_as_errors: true],
+      elixirc_options: [warnings_as_errors: false],
```
Intentional to include? Do we have any tracking issues to be able to restore it in the future?

polvorin replied:
not intentional,  was a leftover :+1:

shanesveller commented on implementations/elixir/ockam/ockam/test/ockam/secure_channel_test.exs:
```diff
@@ -265,18 +245,255 @@ defmodule Ockam.SecureChannel.Tests do
+        encryption_options: [vault: vault]
+      )
+
+    _listener_pid = Ockam.Node.whereis(listener)
```
Do we just want to establish this is a living process and not nil? Can this line be dropped?

polvorin replied:
it's completely useless :+1:     (was a leftover,   wanted to monitor the process and check that it terminated when the connecting side terminates,  but that is done on the spawned receiver,  not this process that is the listener itself)

shanesveller commented on implementations/elixir/ockam/ockam/test/ockam/secure_channel_test.exs:
```diff
@@ -265,18 +245,255 @@ defmodule Ockam.SecureChannel.Tests do
+  end
+
+  test "identity channel inner address is protected" do
+    ## Inner address is the one pointint to the other peer.
```
```suggestion
    ## Inner address is the one pointing to the other peer.
```

shanesveller commented on implementations/elixir/ockam/ockam/test/ockam/secure_channel_test.exs:
```diff
@@ -265,18 +245,255 @@ defmodule Ockam.SecureChannel.Tests do
   end
 
+  test "local secure channel" do
+    {:ok, vault} = SoftwareVault.init()
+    {:ok, alice, alice_id} = Identity.create(@identity_impl)
```
Is it worthwhile to hoist some of these up to setup functions if we're always scrutinizing a pair of Alice/Bob identities in each test body?

polvorin replied:
done.   There is still two "styles" of tests on this file,   one that was here originally and the other for tests moved from the identity secure channel suite.
That can be addressed separately

shanesveller commented on implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/healthcheck.ex:
```diff
@@ -136,6 +138,7 @@ defmodule Ockam.Healthcheck do
          {:worker_init, _worker,
           {:handler_error, reason, _message,
            {Ockam.Transport.TCP, :handle_transport_message, [[]]}}}} ->
+          # How is this defined?
```
👀

hairyhum replied:
This pattern match was taken from the crash report basically.

shanesveller commented on implementations/elixir/ockam/ockam_metrics/test/ockam_metrics/telemetry_poller_test.exs:
```diff
@@ -0,0 +1,111 @@
+  alias Ockam.Vault.Software, as: SoftwareVault
+
+  setup do
+    Ockam.Node.register_address("test")
+    on_exit(fn -> Ockam.Node.unregister_address("test") end)
```
If we use a fixed node name that can collide, should this module be `async: false`?

polvorin replied:
good question :thinking:

shanesveller commented on implementations/elixir/ockam/ockam_metrics/test/test_helper.exs:
```diff
@@ -1 +1,3 @@
-ExUnit.start()
+Application.ensure_all_started(:ockam)
+Application.ensure_all_started(:telemetry)
+ExUnit.start(capture_log: true, trace: true)
```
The scope of impact is limited, but do we want to permanently enable `trace` rather than using the opt-in CLI flag?

polvorin replied:
I don't know what it means :grimacing: ,  copied it to match how it was done on the other libs that already had unit tests defined

shanesveller replied:
Here's the docs from the CLI flag equivalent:

```
  • --trace - runs tests with detailed reporting. Automatically sets
    --max-cases to 1. Note that in trace mode test timeouts will be ignored as
    timeout is set to :infinity
```

It disables any possible parallelism in the test suite as a whole. The order of modules and of test blocks within a given module is non-deterministic without `--seed`, but `mix test --trace` will only ever run a single test module at a time, in some shuffled order of test blocks.

polvorin replied:
created https://github.com/build-trust/ockam/issues/4923  to tackle separately,  whatever we do should be the same on all the libs

hairyhum commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,616 @@
+                +-----------------+                          +-----------------+
+
+
+  The secure channel goes throw three stages:
```
throw -> through

hairyhum commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,616 @@
+    }
+  end
+
+  @spec start_link_channel([initiator_opt()], non_neg_integer) :: :ok | {:error, term()}
```
Can we have `start_link_channel` function return `{:ok, pid}` or `{:ok, pid, info}` for consistency?

polvorin replied:
yes,   I wanted it to be a drop-in replacement for `Ockam.Identity.SecureChannel` without changing its API,   but looking around there is no use of `start_link_channel`  yet :man_shrugging: ,  so ok to change

hairyhum commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,616 @@
+
+  # Check result of the handshake step, send handshake data to the peer if there is a message to exchange,
+  # and possible move to another state
+  defp continue_handshake({:complete, {k1, k2, h, _p}}, state) do
```
Code in this function clause is identical to `defp next_handshake_state({:complete, {k1, k2, h, _p}}, state) do` is there some reason not to reuse it?

hairyhum commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,616 @@
-    case Keyword.get(options, :callback_route) do
-      nil ->
-        {:ok, data}
+  defp handle_outer_message_impl(message, %Channel{channel_state: %Established{} = e} = state) do
```
Does that crash if it's not in the `Established` state?

polvorin replied:
yes.   I'll change it to discard with a loudly error,   if it do receive a message before handshake is complete is very likely an error.   (at that point,   neither   `create_channel`  nor `start_link_channel`  would had returned back to the caller,  nor responder would have contacted his address anywhere).

hairyhum commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,616 @@
+         {:ok, vault} <- vault_from_opts(encryption_options),
+         {:ok, identity} <- identity_from_opts(options),
+         {:ok, key_exchange_state} <-
+           new_xx(vault, Keyword.take(encryption_options, [:static_keypair])) do
```
I know the module is called `XX`, but can we use more descriptive function name here?

hairyhum commented on implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex:
```diff
@@ -1,380 +1,616 @@
+    * IdentityExchange (identity exchange and verification)
+    * Established (channel fully established and peer authenticated)
+
+  At this time, the implementation don't use a proper fsm as that's not directly supported
```
Was the session handshake too complex to accomodate the handshake stage?

polvorin replied:
yes found the session  code (pluggable/etc)  difficult to follow,   mainly what state was passed around,   and what were processes vs what are just callback.    Plus (right now)  handshake spand across several messages,  found it simpler to keep in a single fsm  (kind of like on the previous Channel implementation,  but not like in the Identity channel one)

hairyhum reviewed (commented):
Looks good. A few questions around initialization and error handling.

polvorin committed:
chore(elixir): improve code readability

improve code readability in secure channel implementation, and log
messages received before channel was fully established.

polvorin committed:
chore(elixir): spelling

fix spelling in moduledoc

shanesveller reviewed (approved):
LGTM!
